File: wampmsgscapnp-go/parsers/authenticate.go
package parsers

import (
	"bytes"

	"capnproto.org/go/capnp/v3"

	"github.com/xconnio/wampproto-go/messages"
	"github.com/xconnio/wampproto-messages-capnproto/wampmsgscapnp-go/gen"
)

type Authenticate struct {
	gen *gen.Authenticate
}

func NewAuthenticateFields(g *gen.Authenticate) messages.AuthenticateFields {
	return &Authenticate{gen: g}
}

func (a *Authenticate) Signature() string {
	val, _ := a.gen.Signature()
	return val
}

func (a *Authenticate) Extra() map[string]any {
	return map[string]any{}
}

func AuthenticateToCapnproto(m *messages.Authenticate) ([]byte, error) {
	msg, seg, err := capnp.NewMessage(capnp.SingleSegment(nil))
	if err != nil {
		return nil, err
	}

	auth, err := gen.NewAuthenticate(seg)
	if err != nil {
		return nil, err
	}

	if err = auth.SetSignature(m.Signature()); err != nil {
		return nil, err
	}

	var data bytes.Buffer
	if err := capnp.NewEncoder(&data).Encode(msg); err != nil {
		return nil, err
	}

	return PrependHeader(messages.MessageTypeAuthenticate, &data), nil
}

func CapnprotoToAuthenticate(data []byte) (*messages.Authenticate, error) {
	msg, err := capnp.NewDecoder(bytes.NewReader(data)).Decode()
	if err != nil {
		return nil, err
	}

	auth, err := gen.ReadRootAuthenticate(msg)
	if err != nil {
		return nil, err
	}

	return messages.NewAuthenticateWithFields(NewAuthenticateFields(&auth)), nil
}

File: python/wamp_msgs_capnp/parsers/authenticate.py
import os
from typing import Any
from pathlib import Path

import capnp
from wampproto.messages.authenticate import Authenticate, IAuthenticateFields

from wamp_msgs_capnp.parsers import helpers

root_dir = Path(__file__).resolve().parents[1]
module_file = os.path.join(root_dir, "schemas", "authenticate.capnp")
authenticate_capnp = capnp.load(str(module_file))


class AuthenticateFields(IAuthenticateFields):
    def __init__(self, gen):
        self._gen = gen

    @property
    def signature(self) -> str:
        return self._gen.signature

    @property
    def extra(self) -> dict[str, Any]:
        return {}


def authenticate_to_capnproto(a: Authenticate) -> bytes:
    authenticate = authenticate_capnp.Authenticate.new_message()
    authenticate.signature = a.signature

    data = authenticate.to_bytes_packed()

    return helpers.prepend_header(Authenticate.TYPE, data, b"")


def capnproto_to_authenticate(data: bytes) -> Authenticate:
    message_data, _ = helpers.extract_message(data)
    authenticate_obj = authenticate_capnp.Authenticate.from_bytes_packed(message_data)

    return Authenticate(AuthenticateFields(authenticate_obj))
